### PR TITLE
Pipelines for FBA

### PR DESCRIPTION
**Pull request by yaochaorui**

### CODE REVIEW

File: mmedit/datasets/pipelines/matting_aug.py [hunk 1]
```diff
@@ -555,3 +555,116 @@ def __repr__(self):
                      f'dilate_iter_range={self.dilate_iter_range}, '
                      f'blur_ksizes={self.blur_ksizes})')
         return repr_str
+
+
+@PIPELINES.register_module()
+class CutEdge(object):
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:561]: Remove this

File: mmedit/datasets/pipelines/matting_aug.py [hunk 2]
```diff
@@ -555,3 +555,48 @@ def __repr__(self):
                      f'dilate_iter_range={self.dilate_iter_range}, '
                      f'blur_ksizes={self.blur_ksizes})')
         return repr_str
+
+
+@PIPELINES.register_module()
+class TransformTrimap(object):
+    """Generate two-channel transformed trimap.
+
+    Required key is "trimap", added key is "trimap_o" and "transformed_trimap".
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:564]: remove trimap_o

File: mmedit/datasets/pipelines/matting_aug.py [hunk 3]
```diff
@@ -555,3 +555,48 @@ def __repr__(self):
...
+
+        Args:
+            results (dict): A dict containing the necessary information and
+                data for augmentation.
+
+        Returns:
+            dict: A dict containing the processed data and information.
+        """
+        # generate two channel trimap
+        trimap_o = results['trimap']
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:582]: variable name:
trimap_o -> trimap, for the original trimap
trimap -> trimap2, for 2 channel trimap

File: mmedit/datasets/pipelines/matting_aug.py [hunk 4]
```diff
@@ -555,3 +555,48 @@ def __repr__(self):
...
+        if len(trimap_o.shape) > 2:
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:583]: make this an assertion

File: mmedit/datasets/pipelines/matting_aug.py [hunk 5]
```diff
@@ -555,3 +555,48 @@ def __repr__(self):
...
+            trimap_o = trimap_o.squeeze(2)
+        h, w = trimap_o.shape[:2]
+        trimap = np.zeros((h, w, 2), dtype=np.uint8)
+        trimap[trimap_o == 0, 0] = 255
+        trimap[trimap_o == 255, 1] = 255
+        clicks = np.zeros((h, w, 6))
+        for k in range(2):
+            if (np.count_nonzero(trimap[:, :, k]) > 0):
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:591]:
```python
np.any(trimap[:, :, k])
```

File: mmedit/datasets/pipelines/matting_aug.py [hunk 6]
```diff
@@ -555,3 +555,48 @@ def __repr__(self):
...
+        """
+        # generate two channel trimap
+        trimap_o = results['trimap']
+        if len(trimap_o.shape) > 2:
+            trimap_o = trimap_o.squeeze(2)
+        h, w = trimap_o.shape[:2]
+        trimap = np.zeros((h, w, 2), dtype=np.uint8)
+        trimap[trimap_o == 0, 0] = 255
+        trimap[trimap_o == 255, 1] = 255
+        clicks = np.zeros((h, w, 6))
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:589]: clicks -> trimap_trans

add dtype for zeros

**Comment by innerlee** [comment]
Please add unittests for this function

**Comment by innerlee** [comment]
Add repr

File: mmedit/datasets/pipelines/matting_aug.py [hunk 7]
```diff
@@ -555,3 +555,48 @@ def __repr__(self):
...
+            dict: A dict containing the processed data and information.
+        """
+        trimap = results['trimap']
+        assert len(trimap.shape) == 2
+        h, w = trimap.shape[:2]
+        # generate two-channel trimap
+        trimap2 = np.zeros((h, w, 2), dtype=np.uint8)
+        trimap2[trimap == 0, 0] = 255
+        trimap2[trimap == 255, 1] = 255
+        trimap_trans = np.zeros((h, w, 6), dtype=np.uint8)
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:588]: it is float32

Comment (yaochaorui) [reply mmedit/datasets/pipelines/matting_aug.py]: Fixed. Sorry.

File: mmedit/datasets/pipelines/matting_aug.py [hunk 8]
```diff
@@ -555,3 +559,50 @@ def __repr__(self):
...
+        """
+        trimap = results['trimap']
+        assert len(trimap.shape) == 2
+        h, w = trimap.shape[:2]
+        # generate two-channel trimap
+        trimap2 = np.zeros((h, w, 2), dtype=np.uint8)
+        trimap2[trimap == 0, 0] = 255
+        trimap2[trimap == 255, 1] = 255
+        trimap_trans = np.zeros((h, w, 6), dtype=np.float32)
+        factor = np.array([0.02, 0.08, 0.16])[None, None]
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:593]:
```python
factor = np.array([[[0.02, 0.08, 0.16]]], dtype=np.float32)
```

File: tests/test_loading.py [hunk 9]
```diff
@@ -66,6 +67,21 @@ def test_load_image_from_file():
...
+
+    assert repr(image_loader) == (
+        image_loader.__class__.__name__ +
+        ('(io_backend=disk, key=gt, '
+         'flag=color, save_original_img=False, channel_order=bgr, '
+         'use_cache=True)'))
+    results = image_loader(results)
+    assert results['gt'].shape == (480, 500, 3)
+    np.testing.assert_almost_equal(results['gt'], img_baboon)
+    assert results['gt_path'] == str(path_baboon)
```
Comment (innerlee) [inline tests/test_loading.py:83]: check its cache image_loader.cache is not empty

File: mmedit/datasets/pipelines/matting_aug.py [hunk 10]
```diff
@@ -555,3 +559,50 @@ def __repr__(self):
                      f'dilate_iter_range={self.dilate_iter_range}, '
                      f'blur_ksizes={self.blur_ksizes})')
         return repr_str
+
+
+@PIPELINES.register_module()
+class TransformTrimap:
+    """Generate two-channel transformed trimap.
```
Comment (innerlee) [inline mmedit/datasets/pipelines/matting_aug.py:566]: This is not two channel

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
thanks!

**Pull request merged**